### ISSUE TITLE
Conflicting CB register assignment when using `ParameterBlock` [URGENT]

### ISSUE THREAD

**Issue by nbenty** [shader-slang/slang#306]
How to reproduce:
- Get the latest code from `Falcor/ParameterBlock` branch
- Build `SimpleDeferred` in DebugD3D12
- Place a breakpoint in `SimpleDeferred.cpp`, line 213
- Once you hit that, place a breakpoint in `ProgramReflection.cpp` line 624
When you hit that breakpoint, you'll see that:
- The default block has a CB assigned to space 0, register 0 called 'PerImageCB`
- The `gMaterial` block is getting the same bind location (look at the `mParameterBlocks` member, index 0)

Not sure if I'm doing something wrong or if this is a Slang bug

**Comment by nbenty** [comment]
I inspected Slang's HLSL output, I don't see any trace to `gMaterial` in the shader. The plot thickens...

**Comment by tfoleyNV** [comment]
This sounds like a Slang bug.

**Comment by tfoleyNV** [comment]
I'm hitting an error even earlier, with HLSL compilation failing due to what it thinks are uninitialized variable uses. It is possible that we are running different versions of the HLSL compiler, and hence seeing different behavior. I'll try to debug the issue I'm seeing first (since Slang shouldn't be producing bad output like this).

**Comment by tfoleyNV** [comment]
So, it seems that the compilation error I'm seeing is specific to the SM 5.1 profiles on the version of `fxc` that is on the machine I used. I took the HLSL PS output by Slang and tried compiling it offline (using Slang's `-pass-through fxc` option; I can try with bare fxc if we need). The results on two machines are:

- Machine A
  - `ps_5_1`: compilation fails with (seemingly spurious) `error X4000: variable '...' used without having been completely initialized`
  - `ps_5_0`: compiles without error

- Machine B
  - `ps_5_1`: compiles without error
  - `ps_5_0`: compiles without error

The challenge here is that I honestly have no idea what it is about the generated code that this particular version of `fxc` doesn't like.

For debugging the issue you raised, Nir, I can try to make sure I have a suitable version of the fxc DLL in place so that I can bypass this issue, but it doesn't seem like we can check anything in until we resolve this issue as well.

**Comment by tfoleyNV** [comment]
On the issue I ran into, I've created the following small and pretty clearly correct HLSL program that reproduces the error:

```hlsl
// Problem when using ps_5_1
// Using ps_5_0: works fine

// Only an error when using a specific fxc version
// Other version: work fine

Texture2D ta[3];
SamplerState s;

vector<float,4> eval(
  Texture2D tt,
  float2 uv)
{
  return tt.SampleBias(s, uv, 0.0f);
}

float4 main(
  float2 uv : TEXCOORD
  ) : SV_Target
{
  float4 shAttr[3];

  [unroll]
  for(uint ii = 0; ii < 3; ii++)
  {
    #if 1
    // Leads to `variable 'tt' used without having been completely initialized'
    float4 v = eval(ta[ii], uv);
    #else
    // Not using a subroutine: works fine
    float4 v = ta[ii].SampleBias(s, uv, 0.0f);
    #endif

    shAttr[ii] = v;
  }

  #if 1
  // Bug seems to require that we use only a subset of the computed values
  return shAttr[0];
  #else
  // Result depends on *all* entries: works fine
  return shAttr[0] + shAttr[1] + shAttr[2];
  #endif
}
```

I don't see an easy way for Slang to systematically avoid this problem, so if the version of fxc with this bug is in wide use we'll either need to get people to use a specific version that we bless, or use a different profile (stick with SM 5.0) or a different HLSL compiler (move to `dxc`).

I need to figure out which compiler version is showing the bug (the older or the newer) before I go and bring this to MS.

**Comment by tfoleyNV** [comment]
Okay, I've been able to repro the issue you identify.

Note: the reason you don't see `gMaterial` in the output is because the `LightingPass` code doesn't actually reference it. Variables that are declared in a `.slang` file, but that aren't actually referenced from the chosen entry point, don't get output. I've confirmed that `gMaterial` shows up in the output for `DeferredPass.ps.hlsl`, which seems to be compiling fine.

The error would seem to be either:

- Slang is somehow generating overlapping register assignments. That is entirely possible.

- Slang is generating non-overlapping register assignments, but the Falcor reflection code is forgetting some offsetting math it needs to apply (e.g., maybe the handling of spaces isn't 100% right yet)

I'll try to dig into the first one, since I'd need to rule it out before doing anything on the second one anyway.
